### ISSUE TITLE
False warning on implicit task dependencies (due to overlapping outputs)

### ISSUE THREAD

**Issue by Loffe** [gradle/gradle#22684]
Building the attached gradle project gives false warnings on implicit task dependencies and reused outputs. The output files are located in the same output directory, but with different names.

I'm using a `fileTree()` with a filter to declare the outputs, and the tasks gradle finds conflicting do not have any output file in common.

[gradle-overlapping-outputs-warning.zip](https://github.com/gradle/gradle/files/9989922/gradle-overlapping-outputs-warning.zip)

### Expected Behavior
I expect no deprecation warning.

### Current Behavior
Gradle 7.5 gives a warning on implicit task dependency, and recommends adding `dependsOn` or `mustRunAfter` when there is no such interaction.

```
$ gradle build --console=verbose
> Task :somefile
> Task :assembleDebug

> Task :assembleRelease
Execution optimizations have been disabled for task ':assembleRelease' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: 'C:\Projects\scrap\gradle-overlapping-outputs-warning\build'. Reason: Task ':assembleDebug' uses this output of task ':assembleRelease' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.5.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'C:\Projects\scrap\gradle-overlapping-outputs-warning\build\somefile.txt'. Reason: Task ':assembleRelease' uses this output of task ':assembleDebug' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.5.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.

> Task :build
```

### Context

I have a custom task that uses CMake to compile C++. I have two instances of this task type, one for `Debug` and one for `Release`. In 
FOr one subproject in a multi-project build, we use call out to a CMake project to compile a C++ library.

We have custom tasks for `cmakeGenerate` and `cmakeBuildDebug` and `cmakeBuildRelease`. CMake generates a VisualStudio project in one directory, and the intermediate object files are located within this directory when being built. It is not very easy to separate output directories completely.

In Gradle 6.8.3 I saw no warning on this, but after upgrading to Gradle 7.5 it appeared.

### Steps to Reproduce 
See attached gradle project.

### Your Environment
Gradle 7.5, Windows 10, Java 8

**Comment by jbartok** [comment]
Thank you for providing a valid reproducer.

Issue will be in the backlog of the relevant team but the existence of a workaround makes it non-critical so it might take a while before a fix is made.

***

Gradle 7.x is more strict in detecting of overlapping output of tasks. This is a feature of 7.x, so not a regression.

We believe the below configuration is an issue, because Gradle will only look at the directory of the `FileTree`, not the include/exclude specs. 
```
@OutputFiles
    FileTree getOutputFiles() {
        return project.fileTree(
            dir: project.buildDir,
            include: [
                "${buildType}.txt",
                "${buildType}/*.obj"
            ]
        )
    }
```

**Comment by wolfs** [comment]
The message is not about conflicting outputs. It is about one task (`assembleDebug`) using the outputs of another task (`assembleRelease`) as an input.  If I understand correctly, then input file and the output file of both tasks are in the same directory. The problem here is that Gradle will not take the filtering of outputs into consideration when checking whether they overlap with another input. The reason for that is that the task could produce arbitrary files in the output location already and could overwrite the input of the other task.

Note that Gradle doesn't run the two tasks in parallel either, since it detects that they'll be writing to the same output directory. That did happen for a very long time already.

Like the message says, you should add a `mustRunAfter` between the different build type tasks to choose an order in which you want them to run. That ensures that your build stays reproducible. Another option would be to declare the directory with the object files and the build type txt file as outputs separately.

We know about the use-case to run an external tool (like CMake) on a location which consumes inputs and produces outputs in the same directory. Gradle doesn't currently support this use-case very well and we want to make something about this better in the future. Though we do not have a timeline for this, yet.

**Comment by Loffe** [comment]
Thank you for your response. 

So you are saying using a `fileTree` with includes/excludes as task output is not supported?

If so, I would rather see a warning on that, rather than the current behavior. Especially, since Gradle explicitly lists a file as being used (as input) to another task, when that is not actually the case.

**Comment by wolfs** [comment]
> So you are saying using a fileTree with includes/excludes as task output is not supported?

I am not saying that. A fileTree as an output which selects only part of the outputs of the task in a directory is supported. The meaning currently is: "The task will modify the contents of this directory. After the task finished executing, consider the included files in this file tree as the tasks outputs". This is different to saying: "The task will modify only the files included in this file tree in this directory". This is because there is also the use-case that some tools may produce log files, though you don't want to include the log files in the tasks' output.

**Comment by Loffe** [comment]
Ah, thanks for the explanation, I see the difference. Then I'll add `mustRunAfter` to a number of places, to suppress the warning.
